functorch: fallthrough on calls to custom size/stride/storage_offset calls

bdhirsh opened the pull request:
The problem (that @zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc @zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #109662
* #109023
* #108654
* #109632
* #108656
* #109248
* __->__ #109024

bdhirsh committed:
functorch: fallthrough on calls to custom size/stride/storage_offset calls

[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/DynamicLayer.cpp:
```diff
@@ -499,6 +499,37 @@ TORCH_LIBRARY_IMPL(_, FuncTorchDynamicLayerBackMode, m) {
   m.fallback(torch::CppFunction::makeFromBoxedFunction<&dynamicLayerBackFallback>());
 }
 
+TORCH_LIBRARY_IMPL(aten, FuncTorchDynamicLayerFrontMode, m) {
```
Let me think about it some more. Morally, we shouldn't do the fallthrough business, because one day functorch will just be implemented using pre-dispatch modes and it is possible for each pre-dispatch mode to have their own custom sizes/strides behavior and we don't want to skip all of that logic.

The problem is that functorch can't do a re-dispatch inside of DynamicLayerFrontMode (which is happening because we try to construct a new TensorWrapper, which calls .sym_storage_offset, which is triggering the re-dispatch). So I want to see if we can figure out how to get that to work first

zou3519 replied:
The correct behavior should be: when we are constructing TensorWrapper, we should call .sym_storage_offset at the next mode on the functorch stack.

To get there, I think we just need to ExcludeDispatchKeyGuard on DynamicLayerFrontMode here: https://github.com/pytorch/pytorch/blob/e19a855b4ddead54c26b58d8aa5f8442310bea5e/aten/src/ATen/functorch/ADInterpreters.cpp#L31 instead of doing the fallthrough business

bdhirsh replied:
updated the PR to do as you described (this works)

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -32,6 +32,12 @@ static Tensor materializeGradWrappers(const Tensor& tensor, int64_t current_leve
   if (!tensor.defined()) {
     return tensor;
   }
+  // We don't want to re-enter the DynamicLayerFrontMode **while** we are creating a TensorWrapper.
+  // This can only really happen if we happen to call dispatcher ops while creating the TensorWrapper.
+  // Unfortunately, sym_storage_offset (and friends) are dispatcher ops,
+  // and can enter the dispatcher when our inner tensor is a tensor subclass with custom size/strides
+  // (example: FunctionalTensor)
+  c10::impl::ExcludeDispatchKeyGuard guard(c10::DispatchKey::FuncTorchDynamicLayerFrontMode);
```
A clearer way to phrase this is something like:
- TensorWrapper creation may call dispatcher ops
- We need to ensure that they pass through the functorch stack properly
- In order to do that, we want to call those dispatcher ops at the next layer, hence we disable DynamicLayerFrontMode so the call to the op automatically goes to DynamicLayerBackMode which will then send it to the next layer.

zou3519 reviewed (approved):
LGTM, but please update the comment

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]

bdhirsh committed:
Update on "functorch: fallthrough on calls to custom size/stride/storage_offset calls"

The problem (that zou3519 pointed out) is that functorch assumes that when it create a TensorImpl (like `TensorWrapper`, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/functorch/TensorWrapper.cpp#L43)), it doesn't re-enter the dispatcher.

However, if the inner tensor that we hold is a tensor subclass with custom size/strides, then calls like `sym_storage_offset()` get plumbed to `__torch_dispatch__` as `torch.ops.aten.sym_storage_offset.default`, which is a real op registered to the dispatcher ([here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/runtime/register_prim_ops.cpp#L526)).

cc zou3519, let me know if you think there's a better alternative - in this PR, I just manually audited the ops in that file, and registered fallthroughs to them for the functorch keys (which normally run backend fallbacks).




[ghstack-poisoned]
